File: src/lib/registry.ts
import Database from "better-sqlite3";
import fs from "fs";
import path from "path";
import { closeDbConnection, getDbConnection } from "./db-pool";
import { deleteFilesForDatabase } from "./file-storage";

const DATA_PATH = process.env.DATA_PATH ?? "/data";

let _registry: Database.Database | null = null;

export function getRegistry(): Database.Database {
  if (_registry) return _registry;

  const dbPath = path.join(DATA_PATH, "registry.db");
  fs.mkdirSync(DATA_PATH, { recursive: true });
  _registry = new Database(dbPath);
  _registry.pragma("journal_mode = WAL");
  _registry.exec(`
    CREATE TABLE IF NOT EXISTS databases (
      id             INTEGER PRIMARY KEY AUTOINCREMENT,
      name           TEXT UNIQUE NOT NULL,
      owner          TEXT NOT NULL,
      description    TEXT,
      service_secret TEXT,
      created_at     DATETIME DEFAULT CURRENT_TIMESTAMP,
      status         TEXT DEFAULT 'active'
    );

    CREATE TABLE IF NOT EXISTS file_token_revocations (
      token_id   TEXT PRIMARY KEY,
      db_name    TEXT NOT NULL,
      expires_at DATETIME NOT NULL,
      reason     TEXT,
      revoked_at DATETIME DEFAULT CURRENT_TIMESTAMP
    );

    CREATE INDEX IF NOT EXISTS idx_file_token_revocations_db_name ON file_token_revocations(db_name);
    CREATE INDEX IF NOT EXISTS idx_file_token_revocations_expires_at ON file_token_revocations(expires_at);

    CREATE TABLE IF NOT EXISTS audit_events (
      id         INTEGER PRIMARY KEY AUTOINCREMENT,
      event_type TEXT NOT NULL,
      db_name    TEXT,
      actor      TEXT,
      metadata   TEXT,
      created_at DATETIME DEFAULT CURRENT_TIMESTAMP
    );

    CREATE INDEX IF NOT EXISTS idx_audit_events_created_at ON audit_events(created_at);
    CREATE INDEX IF NOT EXISTS idx_audit_events_db_name ON audit_events(db_name);
    CREATE INDEX IF NOT EXISTS idx_audit_events_event_type ON audit_events(event_type);
  `);
  // Migrate existing installations — add columns if missing
  const cols = _registry
    .prepare("PRAGMA table_info(databases)")
    .all() as { name: string }[];
  if (!cols.some((c) => c.name === "service_secret")) {
    _registry.exec("ALTER TABLE databases ADD COLUMN service_secret TEXT");
  }
  if (!cols.some((c) => c.name === "original_name")) {
    _registry.exec("ALTER TABLE databases ADD COLUMN original_name TEXT");
  }
  if (!cols.some((c) => c.name === "deleted_at")) {
    _registry.exec("ALTER TABLE databases ADD COLUMN deleted_at DATETIME");
  }

  // Auto-provision the control database only when explicitly opted in
  const controlSecret = process.env.CONTROL_DB_SECRET;
  const enableControlDb = (process.env.ENABLE_CONTROL_DB ?? 'false').toLowerCase() === 'true';
  if (enableControlDb && controlSecret) {
    const existing = _registry
      .prepare("SELECT name, service_secret FROM databases WHERE name = 'control'")
      .get() as { name: string; service_secret: string | null } | undefined;
    if (!existing) {
      _registry
        .prepare("INSERT INTO databases (name, owner, description, service_secret) VALUES ('control', 'system', 'Control plane metadata database', ?)")
        .run(controlSecret);
    } else if (!existing.service_secret) {
      _registry
        .prepare("UPDATE databases SET service_secret = ? WHERE name = 'control'")
        .run(controlSecret);
    }

    // Initialise the control database schema directly via better-sqlite3.
    // Running this here (behind ENABLE_CONTROL_DB) guarantees all tables exist
    // before the control plane service starts and tries to use them.
    try {
      const controlDb = getDbConnection('control');
      controlDb.exec(`
        CREATE TABLE IF NOT EXISTS users (
          id          TEXT PRIMARY KEY,
          email       TEXT NOT NULL,
          nube_plan   TEXT NOT NULL DEFAULT 'developer',
          nube_status TEXT NOT NULL DEFAULT 'active',
          created_at  TEXT NOT NULL DEFAULT (datetime('now'))
        );

        CREATE TABLE IF NOT EXISTS instances (
          id                      TEXT PRIMARY KEY,
          url                     TEXT NOT NULL,
          region                  TEXT,
          admin_token             TEXT NOT NULL,
          max_databases           INTEGER NOT NULL DEFAULT 100,
          current_databases_count INTEGER NOT NULL DEFAULT 0,
          is_available            INTEGER NOT NULL DEFAULT 1,
          health_check_at         TEXT,
          created_at              TEXT NOT NULL DEFAULT (datetime('now'))
        );

        CREATE TABLE IF NOT EXISTS databases (
          id                     TEXT PRIMARY KEY,
          user_id                TEXT NOT NULL REFERENCES users(id),
          name                   TEXT NOT NULL,
          display_name           TEXT,
          description            TEXT,
          sqlite_hub_instance_id TEXT NOT NULL REFERENCES instances(id),
          service_secret         TEXT NOT NULL,
          status                 TEXT NOT NULL DEFAULT 'active',
          size_bytes             INTEGER NOT NULL DEFAULT 0,
          created_at             TEXT NOT NULL DEFAULT (datetime('now')),
          updated_at             TEXT
        );

        CREATE TABLE IF NOT EXISTS api_keys (
          id           TEXT PRIMARY KEY,
          user_id      TEXT NOT NULL REFERENCES users(id),
          key_hash     TEXT NOT NULL,
          name         TEXT NOT NULL,
          scope        TEXT NOT NULL DEFAULT 'account',
          created_at   TEXT NOT NULL DEFAULT (datetime('now')),
          last_used_at TEXT,
          status       TEXT NOT NULL DEFAULT 'active'
        );

        CREATE TABLE IF NOT EXISTS usage (
          id               TEXT PRIMARY KEY,
          user_id          TEXT NOT NULL REFERENCES users(id),
          period_year      INTEGER NOT NULL,
          period_month     INTEGER NOT NULL,
          queries_executed INTEGER NOT NULL DEFAULT 0,
          api_calls        INTEGER NOT NULL DEFAULT 0,
          storage_bytes    INTEGER NOT NULL DEFAULT 0,
          calls_success    INTEGER NOT NULL DEFAULT 0,
          calls_client_err INTEGER NOT NULL DEFAULT 0,
          calls_server_err INTEGER NOT NULL DEFAULT 0,
          UNIQUE(user_id, period_year, period_month)
        );

        CREATE TABLE IF NOT EXISTS api_key_databases (
          api_key_id  TEXT NOT NULL REFERENCES api_keys(id),
          database_id TEXT NOT NULL REFERENCES databases(id),
          PRIMARY KEY (api_key_id, database_id)
        );
      `);

      // Seed the default instance row so the control plane can look up
      // this template service automatically on first run.
      // SQLITE_HUB_PUBLIC_URL  — the public URL of this template service (e.g. https://api.mesahub.app)
      // ADMIN_TOKEN            — the admin bearer token for this service
      const instanceUrl  = process.env.SQLITE_HUB_PUBLIC_URL;
      const instanceToken = process.env.ADMIN_TOKEN;
      if (instanceUrl && instanceToken) {
        controlDb
          .prepare(`INSERT OR IGNORE INTO instances (id, url, region, admin_token, max_databases)
                    VALUES ('default', ?, 'default', ?, 100)`)
          .run(instanceUrl, instanceToken);
      }

      console.log('[registry] control database schema initialised');
    } catch (err) {
      console.error('[registry] failed to initialise control database schema:', err);
    }
  }

  return _registry;
}

export interface DbRecord {
  id: number;
  name: string;
  owner: string;
  description: string | null;
  service_secret: string | null;
  created_at: string;
  status: string;
  original_name: string | null;
  deleted_at: string | null;
}

export interface AuditMetrics {
  totalEvents: number;
  eventsLast24h: number;
  byTypeLast24h: Record<string, number>;
}

// Cached prepared statements — created once, reused across requests
let _stmts: {
  list: Database.Statement;
  get: Database.Statement;
  getBySecret: Database.Statement;
  insert: Database.Statement;
  updateSecret: Database.Statement;
} | null = null;

function stmts() {
  if (_stmts) return _stmts;
  const db = getRegistry();
  _stmts = {
    list:         db.prepare("SELECT * FROM databases WHERE status != 'deleted' ORDER BY created_at DESC"),
    get:          db.prepare("SELECT * FROM databases WHERE name = ?"),
    getBySecret:  db.prepare("SELECT * FROM databases WHERE service_secret = ? AND status = 'active'"),
    insert:       db.prepare("INSERT INTO databases (name, owner, description, service_secret) VALUES (?, ?, ?, ?)"),
    updateSecret: db.prepare("UPDATE databases SET service_secret = ? WHERE name = ?"),
  };
  return _stmts;
}

export function listDatabases(): DbRecord[] {
  return stmts().list.all() as DbRecord[];
}

export function getDatabase(name: string): DbRecord | null {
  return (stmts().get.get(name) as DbRecord) ?? null;
}

export function getDbByServiceSecret(secret: string): DbRecord | null {
  return (stmts().getBySecret.get(secret) as DbRecord) ?? null;
}

export function insertDatabase(name: string, owner: string, description?: string, serviceSecret?: string): DbRecord {
  stmts().insert.run(name, owner, description ?? null, serviceSecret ?? null);
  return getDatabase(name)!;
}

export function softDeleteDatabase(name: string): void {
  const epoch = Math.floor(Date.now() / 1000);
  const newName = `${name}-${epoch}`;
  const oldPath = path.join(DATA_PATH, `${name}.db`);
  const newPath = path.join(DATA_PATH, `${newName}.db`);

  // Close open connections before renaming the file
  closeDbConnection(name);

  // Rename DB file + WAL/SHM sidecar files if they exist
  if (fs.existsSync(oldPath)) fs.renameSync(oldPath, newPath);
  if (fs.existsSync(oldPath + "-wal")) fs.renameSync(oldPath + "-wal", newPath + "-wal");
  if (fs.existsSync(oldPath + "-shm")) fs.renameSync(oldPath + "-shm", newPath + "-shm");

  // Remove files owned by this DB during soft-delete to avoid orphaned blobs.
  deleteFilesForDatabase(name);

  getRegistry()
    .prepare(
      "UPDATE databases SET name = ?, original_name = ?, status = 'deleted', deleted_at = datetime('now'), service_secret = NULL WHERE name = ?"
    )
    .run(newName, name, name);
}

export function listDeletedDatabases(): DbRecord[] {
  return getRegistry()
    .prepare("SELECT * FROM databases WHERE status = 'deleted' ORDER BY deleted_at DESC")
    .all() as DbRecord[];
}

export function restoreDatabase(deletedName: string): DbRecord {
  const record = getDatabase(deletedName);
  if (!record || record.status !== "deleted" || !record.original_name) {
    throw new Error("Database not found or not in deleted state");
  }

  const originalName = record.original_name;
  const oldPath = path.join(DATA_PATH, `${deletedName}.db`);
  const newPath = path.join(DATA_PATH, `${originalName}.db`);

  // Rename file back
  if (fs.existsSync(oldPath)) fs.renameSync(oldPath, newPath);
  if (fs.existsSync(oldPath + "-wal")) fs.renameSync(oldPath + "-wal", newPath + "-wal");
  if (fs.existsSync(oldPath + "-shm")) fs.renameSync(oldPath + "-shm", newPath + "-shm");

  getRegistry()
    .prepare(
      "UPDATE databases SET name = ?, original_name = NULL, status = 'active', deleted_at = NULL WHERE name = ?"
    )
    .run(originalName, deletedName);

  return getDatabase(originalName)!;
}

export function hardDeleteDatabase(deletedName: string): void {
  const record = getDatabase(deletedName);
  if (!record || record.status !== "deleted") {
    throw new Error("Database not found or not in deleted state");
  }

  const dbPath = path.join(DATA_PATH, `${deletedName}.db`);

  // Remove DB file + sidecar files
  if (fs.existsSync(dbPath)) fs.unlinkSync(dbPath);
  if (fs.existsSync(dbPath + "-wal")) fs.unlinkSync(dbPath + "-wal");
  if (fs.existsSync(dbPath + "-shm")) fs.unlinkSync(dbPath + "-shm");

  deleteFilesForDatabase(deletedName);

  getRegistry()
    .prepare("DELETE FROM databases WHERE name = ?")
    .run(deletedName);
}

export function updateServiceSecret(name: string, secret: string | null): void {
  stmts().updateSecret.run(secret, name);
}

export function setDatabaseStatus(name: string, status: "active" | "inactive"): void {
  getRegistry().prepare("UPDATE databases SET status = ? WHERE name = ?").run(status, name);
}

export function revokeFileToken(input: {
  tokenId: string;
  dbName: string;
  expiresAt: string;
  reason?: string;
}): void {
  getRegistry()
    .prepare(
      `INSERT INTO file_token_revocations (token_id, db_name, expires_at, reason)
       VALUES (?, ?, ?, ?)
       ON CONFLICT(token_id) DO UPDATE SET
         db_name = excluded.db_name,
         expires_at = excluded.expires_at,
         reason = excluded.reason,
         revoked_at = CURRENT_TIMESTAMP`
    )
    .run(input.tokenId, input.dbName, input.expiresAt, input.reason ?? null);
}

export function isFileTokenRevoked(tokenId: string, dbName: string): boolean {
  const row = getRegistry()
    .prepare(
      `SELECT token_id
       FROM file_token_revocations
       WHERE token_id = ? AND db_name = ? AND datetime(expires_at) > datetime('now')
       LIMIT 1`
    )
    .get(tokenId, dbName) as { token_id: string } | undefined;

  return !!row;
}

export function cleanupExpiredFileTokenRevocations(): { deleted: number } {
  const result = getRegistry()
    .prepare("DELETE FROM file_token_revocations WHERE datetime(expires_at) <= datetime('now')")
    .run();
  return { deleted: result.changes };
}

export function recordAuditEvent(input: {
  eventType: string;
  dbName?: string;
  actor?: string;
  metadata?: Record<string, unknown>;
}): void {
  getRegistry()
    .prepare("INSERT INTO audit_events (event_type, db_name, actor, metadata) VALUES (?, ?, ?, ?)")
    .run(
      input.eventType,
      input.dbName ?? null,
      input.actor ?? null,
      input.metadata ? JSON.stringify(input.metadata) : null
    );
}

export function getAuditMetrics(): AuditMetrics {
  const db = getRegistry();
  const totals = db
    .prepare("SELECT COUNT(*) as total, COALESCE(SUM(CASE WHEN datetime(created_at) >= datetime('now', '-1 day') THEN 1 ELSE 0 END), 0) as last24h FROM audit_events")
    .get() as { total: number; last24h: number };

  const byTypeRows = db
    .prepare(
      `SELECT event_type, COUNT(*) as count
       FROM audit_events
       WHERE datetime(created_at) >= datetime('now', '-1 day')
       GROUP BY event_type`
    )
    .all() as { event_type: string; count: number }[];

  const byTypeLast24h: Record<string, number> = {};
  for (const row of byTypeRows) {
    byTypeLast24h[row.event_type] = row.count;
  }

  return {
    totalEvents: totals.total,
    eventsLast24h: totals.last24h,
    byTypeLast24h,
  };
}

File: server/internal/db/registry.go
// Package db — registry.go manages registry.db, the single source of truth
// for all user databases on this instance. It mirrors registry.ts exactly.
package db

import (
	"database/sql"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"time"
)

// Registry wraps the shared registry.db connection.
type Registry struct {
	db       *sql.DB
	dataPath string
}

// DBRecord mirrors the `databases` table row.
type DBRecord struct {
	ID           int64
	Name         string
	Owner        string
	Description  sql.NullString
	CreatedAt    string
	Status       string
	OriginalName sql.NullString
	DeletedAt    sql.NullString
}

// AuditMetrics holds aggregated audit stats.
type AuditMetrics struct {
	TotalEvents   int64
	EventsLast24h int64
	ByTypeLast24h map[string]int64
}

const registrySchema = `
CREATE TABLE IF NOT EXISTS databases (
  id             INTEGER PRIMARY KEY AUTOINCREMENT,
  name           TEXT UNIQUE NOT NULL,
  owner          TEXT NOT NULL,
  description    TEXT,
  created_at     DATETIME DEFAULT CURRENT_TIMESTAMP,
  status         TEXT NOT NULL DEFAULT 'active',
  original_name  TEXT,
  deleted_at     DATETIME
);

CREATE TABLE IF NOT EXISTS file_token_revocations (
  token_id   TEXT PRIMARY KEY,
  db_name    TEXT NOT NULL,
  expires_at DATETIME NOT NULL,
  reason     TEXT,
  revoked_at DATETIME DEFAULT CURRENT_TIMESTAMP
);
CREATE INDEX IF NOT EXISTS idx_ftr_db_name   ON file_token_revocations(db_name);
CREATE INDEX IF NOT EXISTS idx_ftr_expires   ON file_token_revocations(expires_at);

CREATE TABLE IF NOT EXISTS audit_events (
  id         INTEGER PRIMARY KEY AUTOINCREMENT,
  event_type TEXT NOT NULL,
  db_name    TEXT,
  actor      TEXT,
  metadata   TEXT,
  created_at DATETIME DEFAULT CURRENT_TIMESTAMP
);
CREATE INDEX IF NOT EXISTS idx_audit_created   ON audit_events(created_at);
CREATE INDEX IF NOT EXISTS idx_audit_db_name   ON audit_events(db_name);
CREATE INDEX IF NOT EXISTS idx_audit_event_type ON audit_events(event_type);
`

// OpenRegistry opens (or creates) registry.db at dataPath and applies the
// schema. It is idempotent — safe to call on every startup.
func OpenRegistry(dataPath string) (*Registry, error) {
	if err := os.MkdirAll(dataPath, 0o755); err != nil {
		return nil, fmt.Errorf("registry: mkdir %s: %w", dataPath, err)
	}
	dbPath := filepath.Join(dataPath, "registry.db")
	db, err := open(dbPath)
	if err != nil {
		return nil, fmt.Errorf("registry: open: %w", err)
	}
	if _, err := db.Exec(registrySchema); err != nil {
		return nil, fmt.Errorf("registry: schema: %w", err)
	}
	if err := migrateRegistry(db); err != nil {
		return nil, fmt.Errorf("registry: migrate: %w", err)
	}
	return &Registry{db: db, dataPath: dataPath}, nil
}

// migrateRegistry adds columns introduced after the initial schema without
// dropping data — mirrors the column checks in registry.ts.
func migrateRegistry(db *sql.DB) error {
	type col struct{ name string }
	rows, err := db.Query("PRAGMA table_info(databases)")
	if err != nil {
		return err
	}
	defer rows.Close()

	existing := map[string]bool{}
	for rows.Next() {
		var cid int
		var name, typ string
		var notnull, pk int
		var dflt sql.NullString
		if err := rows.Scan(&cid, &name, &typ, &notnull, &dflt, &pk); err != nil {
			return err
		}
		existing[name] = true
	}
	if err := rows.Err(); err != nil {
		return err
	}

	migrations := []struct {
		col string
		ddl string
	}{
		{"original_name", "ALTER TABLE databases ADD COLUMN original_name TEXT"},
		{"deleted_at", "ALTER TABLE databases ADD COLUMN deleted_at DATETIME"},
	}
	for _, m := range migrations {
		if !existing[m.col] {
			if _, err := db.Exec(m.ddl); err != nil {
				return fmt.Errorf("add column %s: %w", m.col, err)
			}
		}
	}
	return nil
}

// Close shuts down the registry connection.
func (r *Registry) Close() error { return r.db.Close() }

// ── CRUD ─────────────────────────────────────────────────────────────────────

// ListDatabases returns all non-deleted databases ordered by created_at desc.
func (r *Registry) ListDatabases() ([]DBRecord, error) {
	rows, err := r.db.Query(
		`SELECT id, name, owner, description, created_at, status, original_name, deleted_at
		 FROM databases WHERE status != 'deleted' ORDER BY created_at DESC`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	return scanDBRecords(rows)
}

// GetDatabase returns the database record for name (any status).
func (r *Registry) GetDatabase(name string) (*DBRecord, error) {
	row := r.db.QueryRow(
		`SELECT id, name, owner, description, created_at, status, original_name, deleted_at
		 FROM databases WHERE name = ?`, name)
	rec, err := scanDBRecord(row)
	if err == sql.ErrNoRows {
		return nil, nil
	}
	return rec, err
}

// InsertDatabase creates a new database record.
func (r *Registry) InsertDatabase(name, owner string, description *string) (*DBRecord, error) {
	_, err := r.db.Exec(
		`INSERT INTO databases (name, owner, description) VALUES (?, ?, ?)`,
		name, owner, strPtr(description))
	if err != nil {
		return nil, err
	}
	return r.GetDatabase(name)
}

// SetDatabaseStatus updates the status field to 'active' or 'inactive'.
func (r *Registry) SetDatabaseStatus(name, status string) error {
	_, err := r.db.Exec(`UPDATE databases SET status = ? WHERE name = ?`, status, name)
	return err
}

// SoftDeleteDatabase renames the DB file and marks the record as deleted.
// Mirrors softDeleteDatabase() in registry.ts.
func (r *Registry) SoftDeleteDatabase(pool *Pool, name string) error {
	epoch := time.Now().Unix()
	newName := fmt.Sprintf("%s-%d", name, epoch)

	// Rename the DB file (and WAL/SHM sidecars) if they exist.
	oldBase := filepath.Join(r.dataPath, name+".db")
	newBase := filepath.Join(r.dataPath, newName+".db")
	for _, suffix := range []string{"", "-wal", "-shm"} {
		old := oldBase + suffix
		nw := newBase + suffix
		if _, statErr := os.Stat(old); statErr == nil {
			if err := os.Rename(old, nw); err != nil {
				return fmt.Errorf("rename %s: %w", old, err)
			}
		}
	}

	// Close pool connection so WAL is checkpointed before rename.
	pool.mu.Lock()
	if db, ok := pool.conns[name]; ok {
		_ = db.Close()
		delete(pool.conns, name)
	}
	pool.mu.Unlock()

	_, err := r.db.Exec(
		`UPDATE databases
		 SET name = ?, original_name = ?, status = 'deleted', deleted_at = datetime('now')
		 WHERE name = ?`,
		newName, name, name)
	return err
}

// ListDeletedDatabases returns all soft-deleted databases.
func (r *Registry) ListDeletedDatabases() ([]DBRecord, error) {
	rows, err := r.db.Query(
		`SELECT id, name, owner, description, created_at, status, original_name, deleted_at
		 FROM databases WHERE status = 'deleted' ORDER BY deleted_at DESC`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	return scanDBRecords(rows)
}

// RestoreDatabase un-deletes a soft-deleted database.
func (r *Registry) RestoreDatabase(pool *Pool, deletedName string) (*DBRecord, error) {
	rec, err := r.GetDatabase(deletedName)
	if err != nil {
		return nil, err
	}
	if rec == nil || rec.Status != "deleted" || !rec.OriginalName.Valid {
		return nil, fmt.Errorf("database %q not found or not in deleted state", deletedName)
	}

	originalName := rec.OriginalName.String
	oldBase := filepath.Join(r.dataPath, deletedName+".db")
	newBase := filepath.Join(r.dataPath, originalName+".db")
	for _, suffix := range []string{"", "-wal", "-shm"} {
		old := oldBase + suffix
		nw := newBase + suffix
		if _, statErr := os.Stat(old); statErr == nil {
			if err := os.Rename(old, nw); err != nil {
				return nil, fmt.Errorf("rename %s: %w", old, err)
			}
		}
	}

	_, err = r.db.Exec(
		`UPDATE databases SET name = ?, original_name = NULL, status = 'active', deleted_at = NULL WHERE name = ?`,
		originalName, deletedName)
	if err != nil {
		return nil, err
	}
	return r.GetDatabase(originalName)
}

// HardDeleteDatabase permanently removes a soft-deleted database.
func (r *Registry) HardDeleteDatabase(deletedName string) error {
	rec, err := r.GetDatabase(deletedName)
	if err != nil {
		return err
	}
	if rec == nil || rec.Status != "deleted" {
		return fmt.Errorf("database %q not found or not in deleted state", deletedName)
	}

	dbPath := filepath.Join(r.dataPath, deletedName+".db")
	for _, suffix := range []string{"", "-wal", "-shm"} {
		f := dbPath + suffix
		if err := os.Remove(f); err != nil && !os.IsNotExist(err) {
			return fmt.Errorf("remove %s: %w", f, err)
		}
	}

	_, err = r.db.Exec(`DELETE FROM databases WHERE name = ?`, deletedName)
	return err
}

// ── File token revocations ────────────────────────────────────────────────────

// RevokeFileToken inserts or upserts a revocation record for a file access token.
func (r *Registry) RevokeFileToken(tokenID, dbName, expiresAt string, reason *string) error {
	_, err := r.db.Exec(
		`INSERT INTO file_token_revocations (token_id, db_name, expires_at, reason)
		 VALUES (?, ?, ?, ?)
		 ON CONFLICT(token_id) DO UPDATE SET
		   db_name = excluded.db_name,
		   expires_at = excluded.expires_at,
		   reason = excluded.reason,
		   revoked_at = CURRENT_TIMESTAMP`,
		tokenID, dbName, expiresAt, strPtr(reason))
	return err
}

// IsFileTokenRevoked returns true if the token has an active revocation.
func (r *Registry) IsFileTokenRevoked(tokenID, dbName string) (bool, error) {
	var id string
	err := r.db.QueryRow(
		`SELECT token_id FROM file_token_revocations
		 WHERE token_id = ? AND db_name = ? AND datetime(expires_at) > datetime('now')
		 LIMIT 1`,
		tokenID, dbName).Scan(&id)
	if err == sql.ErrNoRows {
		return false, nil
	}
	if err != nil {
		return false, err
	}
	return true, nil
}

// CleanupExpiredTokenRevocations deletes expired revocation records.
func (r *Registry) CleanupExpiredTokenRevocations() (int64, error) {
	res, err := r.db.Exec(
		`DELETE FROM file_token_revocations WHERE datetime(expires_at) <= datetime('now')`)
	if err != nil {
		return 0, err
	}
	return res.RowsAffected()
}

// ── Audit events ──────────────────────────────────────────────────────────────

// RecordAuditEvent appends an audit log entry.
func (r *Registry) RecordAuditEvent(eventType string, dbName, actor *string, metadata map[string]any) error {
	var metaStr *string
	if metadata != nil {
		b, err := json.Marshal(metadata)
		if err != nil {
			return err
		}
		s := string(b)
		metaStr = &s
	}
	_, err := r.db.Exec(
		`INSERT INTO audit_events (event_type, db_name, actor, metadata) VALUES (?, ?, ?, ?)`,
		eventType, strPtr(dbName), strPtr(actor), metaStr)
	return err
}

// GetAuditMetrics returns aggregate audit statistics.
func (r *Registry) GetAuditMetrics() (AuditMetrics, error) {
	var m AuditMetrics
	err := r.db.QueryRow(
		`SELECT COUNT(*),
		        COALESCE(SUM(CASE WHEN datetime(created_at) >= datetime('now', '-1 day') THEN 1 ELSE 0 END), 0)
		 FROM audit_events`).Scan(&m.TotalEvents, &m.EventsLast24h)
	if err != nil {
		return m, err
	}

	rows, err := r.db.Query(
		`SELECT event_type, COUNT(*) FROM audit_events
		 WHERE datetime(created_at) >= datetime('now', '-1 day')
		 GROUP BY event_type`)
	if err != nil {
		return m, err
	}
	defer rows.Close()
	m.ByTypeLast24h = map[string]int64{}
	for rows.Next() {
		var typ string
		var cnt int64
		if err := rows.Scan(&typ, &cnt); err != nil {
			return m, err
		}
		m.ByTypeLast24h[typ] = cnt
	}
	return m, rows.Err()
}

// ── helpers ───────────────────────────────────────────────────────────────────

func strPtr(s *string) any {
	if s == nil {
		return nil
	}
	return *s
}

func scanDBRecord(row *sql.Row) (*DBRecord, error) {
	var r DBRecord
	err := row.Scan(
		&r.ID, &r.Name, &r.Owner, &r.Description,
		&r.CreatedAt, &r.Status, &r.OriginalName, &r.DeletedAt)
	if err != nil {
		return nil, err
	}
	return &r, nil
}

func scanDBRecords(rows *sql.Rows) ([]DBRecord, error) {
	var out []DBRecord
	for rows.Next() {
		var r DBRecord
		if err := rows.Scan(
			&r.ID, &r.Name, &r.Owner, &r.Description,
			&r.CreatedAt, &r.Status, &r.OriginalName, &r.DeletedAt); err != nil {
			return nil, err
		}
		out = append(out, r)
	}
	return out, rows.Err()
}
